### PR TITLE
Value-initialized aggregate and POD

### PR DESCRIPTION
**Pull request by kunjsong01**
Fix for https://github.com/esbmc/esbmc/issues/1006 and https://github.com/esbmc/esbmc/issues/1032. 

Use value initialization for aggregate (including POD).

### CODE REVIEW

**Comment by kunjsong01** [comment]
@rafaelsamenezes Go a couple of ERRORs from the sanitizer build in c2goto clib build step
```
==1829==ERROR: AddressSanitizer: use-after-poison on address 0x62100000e6b0 at pc 0x000000486dd3 bp 0x7ffcb5dfd4b0 sp 0x7ffcb5dfcc70
```
Any idea what's this error about? (and what's this sanitizer build for?)

**Review by lucasccordeiro — APPROVED** [review]

**Comment by fbrausse** [comment]
> Any idea what's this error about? (and what's this sanitizer build for?)

I can't reproduce this sanitizer error locally and the new casts introduced in https://github.com/esbmc/esbmc/pull/1040/commits/a162062d33b25c5d6c736854eba65dd95c956878 where it first started failing don't look suspicious to me at all.

The sanitizer build doesn't run the regression checks, but it does run c2goto as part of the build process. So at least those code paths are being checked with the address sanitizer. Unfortunately, it can't print proper stacktraces and the CI doesn't upload the executable when the test failed, preventing us to generate them manually with e.g. `addr2line`.

**Comment by kunjsong01** [comment]
> > Any idea what's this error about? (and what's this sanitizer build for?)
> 
> I can't reproduce this sanitizer error locally and the new casts introduced in [a162062](https://github.com/esbmc/esbmc/commit/a162062d33b25c5d6c736854eba65dd95c956878) where it first started failing don't look suspicious to me at all.
> 
> The sanitizer build doesn't run the regression checks, but it does run c2goto as part of the build process. So at least those code paths are being checked with the address sanitizer. Unfortunately, it can't print proper stacktraces and the CI doesn't upload the executable when the test failed, preventing us to generate them manually with e.g. `addr2line`.

Neither am I able to reproduce it locally --- the sanitizer build works fine on my PC. 
Just rebased it. So if it's still failing, I'm going to cherry-pick the problematic commit and investigate a bit more locally.

**Comment by fbrausse** [comment]
This may all be bogus, but here's what I got from running the stacktrace addresses through `addr2line` on a local sanitizer build of `c2goto`:
```
WRITE of size 8 at 0x62100000e6b0 thread T0
0x0000000000486dd2: glob at ??:?
0x0000000001661b2d: __bss_start at ??:?
0x000000000154ec7e: __bss_start at ??:?
0x0000000001565353: __bss_start at ??:?
0x0000000000dd81a1: __bss_start at ??:?
0x0000000000b902ca: ?? ??:0
0x0000000000b8cdb7: ?? ??:0
0x00000000009b08e6: ?? ??:0
0x00000000009326eb: std::back_insert_iterator<fmt::v7::detail::buffer<char> > fmt::v7::detail::write_float<std::back_insert_iterator<fmt::v7::detail::buffer<char> >, fmt::v7::detail::dragonbox::decimal_fp<double>, char>(std::back_insert_iterator<fmt::v7::detail::buffer<char> >, fmt::v7::detail::dragonbox::decimal_fp<double> const&, fmt::v7::basic_format_specs<char> const&, fmt::v7::detail::float_specs, char) at /home/kane/scorch/esbmc/debug/_deps/fmt-src/include/fmt/format.h:1802
0x0000000000915722: fmt::v7::detail::buffer_appender<char> fmt::v7::detail::write<char, fmt::v7::detail::buffer_appender<char>, unsigned int, 0>(fmt::v7::detail::buffer_appender<char>, unsigned int) at /home/kane/scorch/esbmc/debug/_deps/fmt-src/include/fmt/format.h:2053
0x00000000008671d2: void std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::_M_realloc_insert<char const (&) [26]>(__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, char const (&) [26]) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/vector.tcc:480
0x0000000000551063: std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::erase(std::_List_const_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_List_const_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/stl_list.h:1556
0x0000000000550798: std::vector<int, std::allocator<int> >* boost::any_cast<std::vector<int, std::allocator<int> > >(boost::any*) at /usr/include/boost/any.hpp:253
0x000000000051d6c3: std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_destroy_node(std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >*) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/stl_tree.h:625
0x00000000005bfa7a: std::_Rb_tree<dstring, std::pair<dstring const, irept>, std::_Select1st<std::pair<dstring const, irept> >, std::less<dstring>, std::allocator<std::pair<dstring const, irept> > >::find(dstring const&) const at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/stl_tree.h:2545
0x000000000051cad9: boost::program_options::error::error(boost::program_options::error const&) at ??:?
0x00007f16a8042082: ?? ??:0
0x00000000004703cd: __asan_stack_free_6 at ??:?
```
and
```
0x62100000e6b0 is located 2480 bytes inside of 4096-byte region [0x62100000dd00,0x62100000ed00)
allocated by thread T0 here:
0x000000000051a0fd: boost::any::holder<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::type() const at /usr/include/boost/any.hpp:208
0x000000000089bbae: std::reference_wrapper<static_lifetime_init(contextt const&, codet&)::{lambda(symbolt const&)#1} const> std::cref<static_lifetime_init(contextt const&, codet&)::{lambda(symbolt const&)#1}>(static_lifetime_init(contextt const&, codet&)::{lambda(symbolt const&)#1} const&) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/refwrap.h:382
0x000000000089bf96: operator() at /home/kane/scorch/esbmc/src/clang-c-frontend/clang_c_main.cpp:35
0x000000000089b84c: void contextt::foreach_operand_in_order<static_lifetime_init(contextt const&, codet&)::{lambda(symbolt const&)#1}>(static_lifetime_init(contextt const&, codet&)::{lambda(symbolt const&)#1}&&) const at /home/kane/scorch/esbmc/src/util/context.h:92
0x0000000001657f5e: __bss_start at ??:?
```

**Comment by kunjsong01** [comment]
Cherry-pick the changes one block at at time, and here's the culprit: 
https://github.com/kunjsong01/esbmc/commit/55e52c4c6d3e08d1c1d8412f562a7c75027f0ed3# 

Sanitizer is not happy about the code in clang::Type::Record case. Comment it out would pass the check.

**Review by lucasccordeiro — APPROVED** [review]
´LGTM.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 1]
```diff
@@ -470,7 +471,12 @@ bool clang_c_convertert::get_var(const clang::VarDecl &vd, exprt &new_expr)
   symbol.file_local = (vd.getStorageClass() == clang::SC_Static) ||
                       (!vd.isExternallyVisible() && !vd.hasGlobalStorage());
 
-  if(symbol.static_lifetime && !symbol.is_extern && !vd.hasInit())
+  bool aggregate_value_init =
+    mode == "C++" ? is_aggregate_type(vd.getType()) : false;
```
Comment (fbrausse) [inline src/clang-c-frontend/clang_c_convert.cpp]: Is checking `mode` required now that `is_aggregate_type` is overridden by the C++ frontend?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 2]
```diff
@@ -1509,3 +1509,40 @@ void clang_cpp_convertert::annotate_ctor_dtor_rtn_type(
...
+  case clang::Type::Elaborated:
+  {
+    const clang::ElaboratedType &et =
+      static_cast<const clang::ElaboratedType &>(the_type);
+    return (is_aggregate_type(et.getNamedType()));
+  }
+  case clang::Type::Record:
+  {
+    const clang::RecordDecl &rd =
+      *(static_cast<const clang::RecordType &>(the_type)).getDecl();
```
Comment (fbrausse) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1535]: I'm assuming the address sanitizer complained here, do you have any insight why that was?

**Review by fbrausse — COMMENTED** [review]
All added regressions check only file-scope aggregate/POD objects. Do they work in function scope as well? What about class scope as members?

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 1 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: spot on. 
not needed any more. Removed.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 2 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1535]: it's actually the if condition below this line. 

>do you have any insight why that was?

nope...

Comment (fbrausse) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1535]: OK, thanks!

**Comment by kunjsong01** [comment]
> Do they work in function scope as well? What about class scope as members?

Function scope - yes. Test cases are added. 
class scope - nope. will raise a ticket and see resolve it afterwards.

**Comment by fbrausse** [comment]
> Function scope - yes. Test cases are added.

Sorry, I meant when the object is defined and initialized within a function. The TCs still have the objects in file scope.

*Edit:* Class can be defined outside, but the object not.

**Comment by kunjsong01** [comment]
> > Function scope - yes. Test cases are added.
> 
> Sorry, I meant when the object is defined and initialized within a function. The TCs still have the objects in file scope.
> 
> _Edit:_ Class can be defined outside, but the object not.

My bad. I forgot to instantiate it within the function scope to make it have local storage. 
So unfortunately it's not working - still calling the ctor rather than being value-inited. 

```python
        do_pod (c:@F@do_pod#):
        // 395 no location
        has_bitfield beans;
        // 396 no location
        FUNCTION_CALL:  has_bitfield(&beans)
        // 397 file 1032_function.cpp line 17 functi
```

raised https://github.com/esbmc/esbmc/issues/1043 for the backlog and removed those invalid TCs.

**Review by fbrausse — APPROVED** [review]
OK, many thanks! LGTM.

**Comment by fbrausse** [comment]
As the address sanitizer now is not checking the C++ frontend when running c2goto in the CI, it might be that this problem will return. Probably best to keep an eye out and/or add a comment to the possibly problematic line in `is_aggregate_type`.

**Pull request merged**